### ISSUE TITLE
problem: pthread_cond_wait doesn't get signaled by ctrl+c

### ISSUE THREAD

**Issue by somdoron** [zeromq/libzmq#3372]
# Issue description
pthread_cond_wait is not being in interrupted by ctrl+c and I cannot stop an app that waits on a thread safe socket.

# Environment

* libzmq version (commit hash if unreleased): master (320741f2c9740832929ec709f18b7ef6f60507c4)
* OS: fedora 29

# Minimal test code / Steps to reproduce the issue

```
int main() {
    zsock_t * client = zsock_new_client (NULL);
    char* str = zsock_str (client);
    puts (str);
    zstr_free (&str);
    return 0;
}

# What's the actual result? (include assertion message & call stack if applicable)

# What's the expected result?
the application should stop after pressing ctrl+c

**Comment by somdoron** [comment]
@sigiesec changed.

If using dealer instead of client the app do exit.

**Comment by somdoron** [comment]
A solution might be, to provide a context function that signal all thread safe sockets.
So a user can call it when interrupted. CZMQ will probably do it by default.

**Comment by bluca** [comment]
can't remember off the top of my head, but we are somehow doing it for the non-safe sockets right?

**Comment by somdoron** [comment]
Also, it might make a performance problem, exiting and entering a mutex...

**Comment by somdoron** [comment]
@bluca, no, the behavior of poll/select and pthread_cond_wait is different.
poll get interrupted and return EINTR, pthread_cond_wait is not.

**Comment by bluca** [comment]
ah true, forgot about that

**Comment by bluca** [comment]
I guess then an API as you suggested is the easiest thing to get it working
